less_slow.py

```diff
@@ -597,6 +615,101 @@ def kernel():
 # ? - Class: 125ns
 # ? - Namedtuple: 183ns
 
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_dict(benchmark):
+    point = {"x": 1.0, "y": 2.0, "flag": True}
+
+    def kernel():
+        return point["x"] + point["y"]
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_class(benchmark):
+    point = PointClass(1.0, 2.0, True)
+
+    def kernel():
+        return point.x + point.y
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_dataclass(benchmark):
+    point = PointDataclass(1.0, 2.0, True)
+
+    def kernel():
+        return point.x + point.y
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_slots_dataclass(benchmark):
+    point = PointSlotsDataclass(1.0, 2.0, True)
+
+    def kernel():
+        return point.x + point.y
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_namedtuple(benchmark):
+    point = PointNamedtuple(1.0, 2.0, True)
+
+    def kernel():
+        return point.x + point.y
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_tuple_indexing(benchmark):
+    point = (1.0, 2.0, True)
+
+    def kernel():
+        return point[0] + point[1]
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+@pytest.mark.benchmark(group="composite-structs-access")
+def test_structs_access_tuple_unpacking(benchmark):
+    x, y, _ = (1.0, 2.0, True)
+
+    def kernel():
+        return x + y
+
+    result = benchmark(kernel)
+    assert result == 3.0
+
+
+# ? When accessing pre-constructed objects, the performance gap narrows
```

Comment:
I am afraid this benchmarking tool isn’t accurate to a nano-second resolution. It’s wiser to repeat more member accesses inside the kernel function for higher precision.

Reply:
Yes, I think I'll keep experimenting with better ways to benchmark the attribute access to get more reliable timings, then open a separate PR just for that.